ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/HistoricityCheck.hs

```diff
@@ -0,0 +1,163 @@
+-- and @MsgAwaitReply@s are historical.
+data HistoricityCheck m blk = HistoricityCheck {
+    -- | Determine whether the received @MsgRollBackward@ is historical. The
+    -- input is the oldest state that was rolled back. (Note that rollbacks of
+    -- depth zero are hence never historical.).
+    judgeMessageHistoricity ::
+         HistoricalChainSyncMessage
+      -> HeaderStateWithTime blk
+      -> m (Either HistoricityException ())
+  }
+
+-- | ChainSync historicity checks are performed for @MsgRollBackward@s and
+-- @MsgAwaitReply@s, see 'HistoricityCheck'.
+data HistoricalChainSyncMessage =
+    HistoricalMsgRollBackward
+  | HistoricalMsgAwaitReply
+  deriving stock (Show, Eq)
+
+data HistoricityException =
+  -- | We received a @MsgRollBackward@ or a @MsgAwaitReply@ at a time too early
+  -- for it to plausibly have been sent by an honest caught-up peer.
+  --
+  -- INVARIANT: @'historicityCutoff' < 'arrivalTime' `diffRelTime` 'slotTime'@
+  forall blk. HasHeader blk => HistoricityException {
+      historicalMessage :: HistoricalChainSyncMessage
+      -- | Depending on 'historicalMessage':
+      --
+      --  * 'HistoricalMsgRollBackward': The oldest header that was rewound.
+      --
+      --  * 'HistoricalMsgAwaitReply': The tip of the candidate fragment.
+    , historicalPoint   :: !(Point blk)
+      -- | The time corresponding to the slot of the oldest header that was
+      -- rolled back.
+    , slotTime          :: !RelativeTime
+      -- | When the @MsgRollBackward@ was received.
+    , arrivalTime       :: !RelativeTime
+    , historicityCutoff :: !HistoricityCutoff
+    }
+  deriving anyclass (Exception)
+
+deriving stock instance Show HistoricityException
+
+instance Eq HistoricityException where
+  (==)
+    (HistoricityException l0 (l1 :: Point l) l2 l3 l4)
+    (HistoricityException r0 (r1 :: Point r) r2 r3 r4)
+    = case eqT @l @r of
+        Nothing   -> False
+        Just Refl -> (l0, l1, l2, l3, l4) == (r0, r1, r2, r3, r4)
+
+-- ^ The maximum age of a @MsgRollBackward@ or @MsgAwaitReply@ at arrival time,
+-- constraining the age of the oldest rewound header or the tip of the candidate
+-- fragment, respectively.
+--
+-- This should be set to at least the maximum duration (across all eras) of a
+-- stability window (the number of slots in which at least @k@ blocks are
+-- guaranteed to arise).
+--
+-- For example, on Cardano mainnet today, the Praos Chain Growth property
+-- implies that @3k/f@ (=129600) slots (=36 hours) will contain at least @k@
+-- (=2160) blocks. (Byron has a smaller stability window, namely @2k@ (=24 hours
+-- as the Byron slot length is 20s). Thus a peer rolling back a header that is
+-- older than 36 hours or signals that it doesn't have more headers is either
+-- violating the maximum rollback or else isn't a caught-up node. Either way, a
+-- syncing node should not be connected to that peer.
+newtype HistoricityCutoff = HistoricityCutoff {
```

Comment:
To reviewers: Please say if you don't like this name. A more descriptive but more mouthful alternative would be `MaxRollbackAndAwaitReplyAge` for example.

Reply:
Yeah, "historicity" might be too generic, but maybe once one is familiar with its meaning in this context that won't be an issue anymore. I'm perfectly ok with the longer alternative, and I'd prefer it if it'd be up to me, but I'm fine either way.

Reply:
I think "Historicity" strikes a good balance.